### PR TITLE
feature[v2]: enabling authentication for metric api scaler

### LINKED ISSUE #1082: Support basic authentication for Metrics API scaler
### Use-Case

Support basic authentication for Metrics API scaler

### Reference 
https://github.com/kedacore/keda/issues/929

### PR DESCRIPTION
**Pull request by aman-bansal**
Signed-off-by: aman-bansal <bansalaman2905@gmail.com>

This PR is to add authentication for metric_api_scaler. Three methods have been added. They will be identified on the basis of `authMode` parameter.

1. API Key Based auth: auth mode would `apiKeyAuth`. API key needs to be provided. Default behaviour would be to send it in headers. This can be changed via method and header/query param keys can be changed via keyParamName.

2. Basic Auth: auth mode would `basicAuth`. Username is compulsory. Password isn't. This is based on experience over basic auth. Some people just put api key as the username to ease out authentication at their end.

3. Client Certification: auth mode would `tlsAuth`. cert, cacert and key needs to be provided.

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO)
- [x] Tests have been added
- [x] A PR is opened to update the documentation on https://github.com/kedacore/keda-docs
- [x] Changelog has been updated

Relates to #1082 #1083 #1084 

[Edit]
Docs PR attached https://github.com/kedacore/keda-docs/pull/260

### CODE REVIEW

**Comment by aman-bansal** [comment]
@tomkerkhove please review this PR and do suggest if I missed something. This will enable metric api scaler to authentication via api key, basic auth, and client certificates. In the meantime, I am working on docs.

File: pkg/scalers/metrics_api_scaler.go
Comment (turbaszek) [inline pkg/scalers/metrics_api_scaler.go:122]: How about using negation in this if and then returning instantly?
```
if val, ok := authParams["authMode"]; !ok {
    return here
}
```
In this way we will limit the indentation

File: pkg/scalers/tls_config.go
Comment (zroubalik) [inline pkg/scalers/tls_config.go:1]: Could you please move this out of the `scalers` package? `pkg/util` or `pkg/util/tls` etc

File: CHANGELOG.md
Comment (zroubalik) [inline CHANGELOG.md:45]: Since we are adding Metrics API Scaler in the same release (2.0), I wouldn't make a separate entry in the Changelog, but you can add a link to your PR to the extisting Metrics API Scaler entry

**Review by zroubalik — COMMENTED** [review]
Thanks, minor nits

File: CHANGELOG.md
Comment (aman-bansal) [reply CHANGELOG.md]: yes, this makes sense. I missed the new section. Thanks. I have removed it.

File: pkg/scalers/tls_config.go
Comment (aman-bansal) [reply pkg/scalers/tls_config.go]: This is done. Since this is generic, moving this to util will help to manage other scalers better. Thanks!

File: pkg/scalers/metrics_api_scaler.go
Comment (aman-bansal) [reply pkg/scalers/metrics_api_scaler.go]: oh,  good catch. looks much cleaner now.

**Comment by turbaszek** [comment]
@aman-bansal have you tested it locally? If not, I can help over the weekend or you may find this useful: https://github.com/turbaszek/keda-example/pull/1 (sorry for no docs)

**Comment by zroubalik** [comment]
> @aman-bansal have you tested it locally? If not, I can help over the weekend or you may find this useful: [turbaszek/keda-example#1](https://github.com/turbaszek/keda-example/pull/1) (sorry for no docs)

We should have e2e test for this scaler probably

**Comment by aman-bansal** [comment]
> @aman-bansal have you tested it locally? If not, I can help over the weekend or you may find this useful: [turbaszek/keda-example#1](https://github.com/turbaszek/keda-example/pull/1) (sorry for no docs)

This is great! I was actually looking for ways to test this. I'll try to create one sample for each authorization. If there is any doubt I will raise in this thread itself.

**Comment by aman-bansal** [comment]
> > @aman-bansal have you tested it locally? If not, I can help over the weekend or you may find this useful: [turbaszek/keda-example#1](https://github.com/turbaszek/keda-example/pull/1) (sorry for no docs)
> 
> We should have e2e test for this scaler probably

This will be tricky. We have to add one sample app which can provide support for each authorization method. Will need a bit help in this case.

**Comment by turbaszek** [comment]
> We should have e2e test for this scaler probably

@zroubalik I agree that e2e would be awesome. Do we have any documentation on how we approach such testing? 

> This will be tricky. We have to add one sample app which can provide support for each authorization method. Will need a bit help in this case.

I would say that one app with 4 endpoints, each with different auth methods should serve the purpose. And I'm happy to help with doing this :)

File: pkg/scalers/metrics_api_scaler.go
Comment (turbaszek) [inline pkg/scalers/metrics_api_scaler.go:132]:
```suggestion
	val, ok := authParams["authMode"]
	if !ok {
		return &meta, nil
	}
```
How about consolidating this? Probably it would be also reasonable to rename `val` to something like `authParams`

Comment (turbaszek) [reply pkg/scalers/metrics_api_scaler.go:132]: Especially that we are using `authParams["apiKey"]` multiple times in next lines

**Comment by aman-bansal** [comment]
Took time but finally, build HTTP authentication example, I will add the documentation in some time. But do check if you get the time. https://github.com/aman-bansal/keda-http-auth-example

File: pkg/scalers/metrics_api_scaler.go
Comment (turbaszek) [inline pkg/scalers/metrics_api_scaler.go:45]:
```suggestion
	password string // optional
```

File: pkg/scalers/metrics_api_scaler.go
Comment (turbaszek) [inline pkg/scalers/metrics_api_scaler.go:36]:
```suggestion
	method method // +default is header
```

File: pkg/scalers/metrics_api_scaler.go
Comment (turbaszek) [inline pkg/scalers/metrics_api_scaler.go:62]: Personally I would use just `string` as for me this new type creates some confusion

File: pkg/scalers/metrics_api_scaler.go
Comment (turbaszek) [inline pkg/scalers/metrics_api_scaler.go:88]:
```suggestion
				Timeout:   3 * time.Second,
```
How about creating `defaultTimeout` variable that will be used here and in L97?

File: pkg/scalers/metrics_api_scaler.go
Comment (turbaszek) [inline pkg/scalers/metrics_api_scaler.go:135]:
```suggestion
		if len(authParams["apiKey"]) == 0 {
```
I think this may result in an error as the key may not exists

File: pkg/scalers/metrics_api_scaler.go
Comment (turbaszek) [inline pkg/scalers/metrics_api_scaler.go:278]: How about using switch expression here? It may be easier to extend it in future if we decide to add additional auth method :)

File: pkg/scalers/metrics_api_scaler.go
Comment (turbaszek) [inline pkg/scalers/metrics_api_scaler.go:305]: Is it possible that we will reach this place? With `meta.enableAPIKeyAuth` and both above `if` we may need `else` caluse?

File: pkg/scalers/metrics_api_scaler.go
Comment (turbaszek) [reply pkg/scalers/metrics_api_scaler.go:36]: This gives a better feeling what parameters belong to single "section"

File: pkg/scalers/metrics_api_scaler.go
Comment (aman-bansal) [reply pkg/scalers/metrics_api_scaler.go:135]: I am sensing your concern is with len function. But this will not fail. authParams is a map of string. string can never be nil in go, that's why len will never fail.

File: pkg/scalers/metrics_api_scaler.go
Comment (aman-bansal) [reply pkg/scalers/metrics_api_scaler.go:305]: we'll never reach here. the default method is the header. So always one condition is true. I used else if to have better clarity of why we are doing this.

File: pkg/scalers/metrics_api_scaler.go
Comment (turbaszek) [reply pkg/scalers/metrics_api_scaler.go:135]: Oh, interesting! I didn't know. No objections then 🚀

File: pkg/scalers/metrics_api_scaler.go
Comment (turbaszek) [inline pkg/scalers/metrics_api_scaler.go:66]:
```suggestion
	headerMethod     method = "header"
	queryParamMethod        = "query"
```
How about this?  I think this may introduce some clarity

File: pkg/scalers/metrics_api_scaler.go
Comment (turbaszek) [reply pkg/scalers/metrics_api_scaler.go:305]: I would do `if { } else { }` with a comment in else about fallback to default option. We have to remember that someone will look this code in future and will have to figure out it the code will reach here or no :) 

Btw. I personally I prefere `meta.method == queryParam` as this is more natural to read "(provided) meta method is queal to queryParam"

File: pkg/scalers/metrics_api_scaler.go
Comment (aman-bansal) [reply pkg/scalers/metrics_api_scaler.go]: makes sense. done!

File: pkg/scalers/metrics_api_scaler.go
Comment (aman-bansal) [reply pkg/scalers/metrics_api_scaler.go]: makes sense. adding a new variable with small cardinality adds a bit of confusion. made it string. will add more details in the documentation to have concretely acceptable values

File: pkg/scalers/metrics_api_scaler.go
Comment (aman-bansal) [reply pkg/scalers/metrics_api_scaler.go:271]: Yes, I am already seeing one more use case about JWT token. Completely makes sense. Thanks!

File: pkg/scalers/metrics_api_scaler.go
Comment (aman-bansal) [reply pkg/scalers/metrics_api_scaler.go]: I have removed this section altogether. This will now be part of the documentation itself.

File: pkg/scalers/metrics_api_scaler.go
Comment (aman-bansal) [reply pkg/scalers/metrics_api_scaler.go:298]: Right! Now it will be easier for the third party user to actually read the flow with more clarity. He doesn't have to retrace the whole code sequence to understand this section.

File: pkg/scalers/metrics_api_scaler.go
Comment (aman-bansal) [reply pkg/scalers/metrics_api_scaler.go:80]: Whats L97? Sorry but I am not aware of this.

File: pkg/scalers/metrics_api_scaler.go
Comment (turbaszek) [inline pkg/scalers/metrics_api_scaler.go:37]:
```suggestion
	method string // way of providing auth key, either "header" using X-API-KEY header or "query" by providing api_key in query params
```
This probably will be more detailed in the docs but I think it's still a good idea to have good clarification in code

File: pkg/scalers/metrics_api_scaler.go
Comment (turbaszek) [reply pkg/scalers/metrics_api_scaler.go:80]: No it is Line 89, when we return other Client:
```

  	return &metricsAPIScaler{
 		metadata: meta,
 		client: &http.Client{
 			Timeout: 3 * time.Second,
 		},
 	}, nil
```
I think it makes sense to have a default value for parameters that are used in two places

File: pkg/scalers/metrics_api_scaler.go
Comment (turbaszek) [inline pkg/scalers/metrics_api_scaler.go:39]:
```suggestion
	method string // way of providing auth key, either "header" or "query"
	// keyParamName  is either header key or query param used for passing auth token
	// default header is X-API-KEY, defaul query param is api_key
	keyParamName string
	apiKey       string
```
What do you think?

File: pkg/scalers/metrics_api_scaler.go
Comment (aman-bansal) [reply pkg/scalers/metrics_api_scaler.go]: oh got it :).

File: pkg/scalers/metrics_api_scaler.go
Comment (aman-bansal) [reply pkg/scalers/metrics_api_scaler.go]: more clear. Thanks!

File: pkg/scalers/metrics_api_scaler.go
Comment (turbaszek) [inline pkg/scalers/metrics_api_scaler.go:60]: I'm not 100% but I think we should add `authenticationType` to two other methods, otherwise go will resolve the type from assigned value?

File: pkg/scalers/metrics_api_scaler.go
Comment (turbaszek) [inline pkg/scalers/metrics_api_scaler.go:35]:
```suggestion
	method           string // way of providing auth key, either "header" (default) or "query"
```

File: pkg/scalers/metrics_api_scaler.go
Comment (turbaszek) [inline pkg/scalers/metrics_api_scaler.go:93]:
```suggestion
        client := &http.Client{
            Timeout:   defaultTimeOut,
        }
        
	if meta.enableTLS {
		config, err := kedautil.NewTLSConfig(meta.cert, meta.key, meta.ca)
		if err != nil {
			return nil, err
		}
		client.Transport = &http.Transport{TLSClientConfig: config}
	}

	return client, nil
```
I think may simplifies things a bit

**Review by turbaszek — APPROVED** [review]
Few nits but otherwise it looks good to me! 👏

File: pkg/scalers/metrics_api_scaler.go
Comment (aman-bansal) [reply pkg/scalers/metrics_api_scaler.go]: oh didn't know that automatic type assignment works for zero indexed type declaration i.e. with iota. In testing, this didn't get caught because of its string representation.

**Comment by aman-bansal** [comment]
> Few nits but otherwise it looks good to me! 👏

Thank you so much @turbaszek. Great learning. Will be more careful with future PRs. :)

File: pkg/util/tls_config.go
Comment (turbaszek) [inline pkg/util/tls_config.go:29]: According to code scanning
> InsecureSkipVerify should not be used in production code.

@tomkerkhove how should we address this feedback?

Comment (ahmelsayed) [reply pkg/util/tls_config.go:29]: I'm not sure why we need to skip verification if we're setting the RootCAs on the config. I haven't tried it, but I'd think it shouldn't be needed. Maybe we just need to append the system ca certs first 

```go
caCertPool, err := x509.SystemCertPool()
if err != nil {
    ....
}
caCertPool.AppendCertsFromPEM([]byte(caCert))
...
```
[source](https://stackoverflow.com/q/62718784)

Comment (turbaszek) [reply pkg/util/tls_config.go:29]: @aman-bansal what do you think?

**Comment by turbaszek** [comment]
@tomkerkhove @zroubalik is there any blocker to merging this PR other than rebasing it?

**Comment by tomkerkhove** [comment]
I'll leave it up to @ahmelsayed @zroubalik for reviewing this

File: pkg/util/tls_config.go
Comment (aman-bansal) [reply pkg/util/tls_config.go:29]: Hey, I am back finally.
So I believe we shouldn't use this because keda when running in the production systems this could lead to issues. But should we make this configurable? 

Secondly, I believe the system ca cert is not required for this. We need system ca cert only if using self-signed cert. I believe we should let the clients specify if they need this via configuration. Because some clients might need this.

**Comment by aman-bansal** [comment]
@ahmelsayed @zroubalik can we have a final review over this. I am back in 100% capacity and would love to complete this so that I can pick something next :).

File: pkg/util/tls_config.go
Comment (zroubalik) [reply pkg/util/tls_config.go:29]: @aman-bansal I agree

**Comment by zroubalik** [comment]
@aman-bansal could you please rebase this PR, there's a conflict. I think that we are good to go with the merge and then we can address the `config.InsecureSkipVerify = true` problem in a separate issue.

**Comment by aman-bansal** [comment]
@zroubalik @ahmelsayed Merge conflicts have been resolved.

**Review by zroubalik — APPROVED** [review]
LGTM thanks @aman-bansal !

**Pull request merged**
